### PR TITLE
LWG-3036 polymorphic_allocator::destroy is extraneous

### LINKED ISSUE #753: <memory_resource>: std::pmr::polymorphic_allocator<T>::destroy missing
https://en.cppreference.com/w/cpp/memory/polymorphic_allocator/destroy
http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2017/n4713.pdf §23.12.3.2 #14
Seems to be missing.
https://github.com/microsoft/STL/blob/master/stl/inc/xpolymorphic_allocator.h#L190
https://godbolt.org/z/iFfpPT

I had expected something like:

        template<class _Ty>
        void destroy(_Ty* _Ptr) {
            // destroys _Ptr as if by _Ptr->~_Ty().
            _Ptr->~_Ty();
        }

If this is indeed so, I can provide a PR.

Also tracked by DevCom-394908 and Microsoft-internal VSO-737704 / [AB#737704](https://devdiv.visualstudio.com/0bdbc590-a062-4c3f-b0f6-9383f67865ee/_workitems/edit/737704).

### LINKED ISSUE #1454: LWG-3036 polymorphic_allocator::destroy is extraneous
LWG-3036 `polymorphic_allocator::destroy` is extraneous

### PR DESCRIPTION
**Pull request by CaseyCarter**
By defining `polymorphic_allocator::destroy` - despite that it's equivalent to the default in `allocator_traits` - as deprecated. Also hack up the internal `_Has_no_alloc_destroy` trait so our optimization that avoids calling `destroy` on trivially destructible types keeps working with `polymorphic_allocator`.

Fixes #1454 and fixes #753.

### CODE REVIEW

File: stl/inc/xpolymorphic_allocator.h [hunk 1]
```diff
@@ -274,6 +273,11 @@ namespace pmr {
             _Uses_allocator_construct(_Ptr, _Al, *this, _STD forward<_Types>(_Args)...);
         }
 
+        template <class _Uty>
+        _CXX20_DEPRECATE_POLYMORPHIC_ALLOCATOR_DESTROY void destroy(_Uty* const _Ptr) noexcept /* strengthened */ {
```
Comment (miscco) [inline stl/inc/xpolymorphic_allocator.h:277]: Did you actually implement it to deprecate it?

Comment (CaseyCarter) [reply stl/inc/xpolymorphic_allocator.h:277]: Yes. After holding out hope forever that LWG would just strike the silly thing.

File: stl/inc/yvals_core.h [hunk 2]
```diff
@@ -1007,7 +1007,19 @@
 #define _CXX20_DEPRECATE_MOVE_ITERATOR_ARROW
 #endif // ^^^ warning disabled ^^^
 
-// next warning number: STL4032
+#if _HAS_CXX20 && !defined(_SILENCE_CXX20_POLYMORPHIC_ALLOCATOR_DESTROY_DEPRECATION_WARNING) \
```
Comment (CaseyCarter) [inline stl/inc/yvals_core.h:1010]: This deprecation needs to be active in C++17 mode: we don't want people to start writing code that uses `polymorphic_allocator::destroy` after applying this PR that deprecates it.

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
Looks good to me, thanks for implementing this (especially the `_Has_no_alloc_destroy` optimization). I pushed a few changes:

* Explain that this "is deprecated in C++17 by LWG-3036" and change the macros to `CXX17`.
* Add a space at the end of the sentence mentioning `allocator_traits::destroy` as an alternative. Add another in the `warning STL4021` message above.
* Add parentheses after `destroy_at` and `allocator_traits::destroy` to match the style used in other deprecation warning messages.
* In test code, mark the constructor `S(bool&)` as `explicit`.

**Review by cbezault — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
Thanks again for implementing this user-requested member function, deprecating it so users will decide they don't want it after all :joy_cat:, and ensuring that the STL continues to generate efficient code! :rocket:

**Pull request merged**